projects/hathor-wallet-headless/hsm-integration.md

```diff
@@ -0,0 +1,280 @@
+- /wallet/simple-send-tx
+- /wallet/send-tx
+- /wallet/utxo-consolidation
+- /wallet/create-token
+- /wallet/mint-tokens
+- /wallet/melt-tokens
+- /wallet/create-nft
+
+Comparing these with the direct usage of the `prepareTransaction` method, we can see that this method is called by all of the routes above:
+- [prepareCreateNewToken](https://github.com/HathorNetwork/hathor-wallet-lib/blob/a0e20f6aa9c07461d7784e3133a6fedc4b72fabf/src/new/wallet.js#L1487-L1494)
+- [prepareDelegateAuthorityData](https://github.com/HathorNetwork/hathor-wallet-lib/blob/a0e20f6aa9c07461d7784e3133a6fedc4b72fabf/src/new/wallet.js#L1878)
+- [prepareDestroyAuthorityData](https://github.com/HathorNetwork/hathor-wallet-lib/blob/a0e20f6aa9c07461d7784e3133a6fedc4b72fabf/src/new/wallet.js#L1957)
+- [prepareMeltTokensData](https://github.com/HathorNetwork/hathor-wallet-lib/blob/a0e20f6aa9c07461d7784e3133a6fedc4b72fabf/src/new/wallet.js#L1791-L1798)
+- [prepareMintTokensData](https://github.com/HathorNetwork/hathor-wallet-lib/blob/a0e20f6aa9c07461d7784e3133a6fedc4b72fabf/src/new/wallet.js#L1683-L1690)
+- [new/transaction.js/prepareTx](https://github.com/HathorNetwork/hathor-wallet-lib/blob/f211d0167748e5aa5bb674e65c1e927aeb1838e2/src/new/sendTransaction.ts#L253)
+  - called by [SendTransaction.run](https://github.com/HathorNetwork/hathor-wallet-lib/blob/f211d0167748e5aa5bb674e65c1e927aeb1838e2/src/new/sendTransaction.ts#L479)
+  - called by [sendManyOutputsTransaction](https://github.com/HathorNetwork/hathor-wallet-lib/blob/a0e20f6aa9c07461d7784e3133a6fedc4b72fabf/src/new/wallet.js#L1235)
+  - called by:
+    - [consolidateUtxos](https://github.com/HathorNetwork/hathor-wallet-lib/blob/a0e20f6aa9c07461d7784e3133a6fedc4b72fabf/src/new/wallet.js#L1045)
+    - [sendTransaction](https://github.com/HathorNetwork/hathor-wallet-lib/blob/a0e20f6aa9c07461d7784e3133a6fedc4b72fabf/src/new/wallet.js#L1185)
+
+A new function could be added to the wallet storage, allowing the `prepareTransaction` to identify the wallet is linked to an HSM device. From this point, the function itself could retrieve the connection credentials from the config file and call the HSM lib directly, executing the signature.
+
+Extra attention would be needed to ensure that thrown exceptions from the HSM lib would be correctly handled by the Wallet Lib, and that the headless end user would receive a friendly error message.
+
+## Avoiding the use of the `xPub`
+There is the possibility of some partner needing a wallet with a higher level of security, avoiding even the export of the `xPub` data.
+
+Should this become a requirement, another path to implement this integration would be to modify the Wallet Lib, allowing the starting of a wallet without a way to derive public keys internally.
```

Comment:
The wallet nowadays is ready to start a read only wallet from a list of addresses only. I think this would be enough for this requirement, we would just need to implement this support in the headless, not in the lib.

Reply:
Reworded this section on [66c6c50](https://github.com/HathorNetwork/rfcs/pull/67/commits/66c6c50949a9f15e81867c416b5dca57885af1bb). However, I could not find when did this feature was implemented to add as a reference on the document. Do you know which PR added it?

Reply:
This was a long time ago. You can reference the line with the attribute of the constructor in the `HathorWallet` class.

https://github.com/HathorNetwork/hathor-wallet-lib/blob/v1.1.0/src/new/wallet.js#L107

Reply:
I incorporated our ideas on https://github.com/HathorNetwork/hathor-wallet-headless/issues/393 and made a reference to them on the design on commit [c8a5afd](https://github.com/HathorNetwork/rfcs/pull/67/commits/c8a5afd14886ae8f453c17914f5602279cba89ac)